### DISCUSSION TITLE
Quasar Dialog - v-close-popup not working after await?

### DISCUSSION THREAD

**Discussion by sonicviz** [quasarframework/quasar#16941] [General - Components / Directives / etc] [answered]
Hi, 
I have  a dialog set up in the template: 
`      <q-dialog v-model="saveFile" persistent>...`
with some inputs and buttons.

Triggered by this button:
`        <q-btn class="col-12 q-mt-sm text-grey-1" label="Save" color="primary" @click="saveFile = true" v-if="showSave" />`

The dialog has a cancel and add product buttons, with validation checks.
```
          <q-card-actions align="right" class="text-primary">
            <q-btn flat label="Cancel" v-close-popup />
            <q-btn flat label="Add product" v-close-popup="closeSave" @click="saveProduct" />
          </q-card-actions>
```

After validation is correct the v-close-popup="closeSave" is used to trigger the closing of the popup.       

It works fine with synchronous code, but if I have an await method it doesn't close, despite closeSave.value = true
It needs await methods as the saveProduct function does some api calls.

I can replicate the behavior with a timeout in the saveProduct function as well (just to confirm it happens in different async contexts):

      setTimeout(() => {
        AlertTypes.value = "bg-green text-white";
        message.value = "Product saved.";
        quasar.loading.hide();
        closeSave.value = true;
      }, 3000);
      
In this case the dialog will stay open as well. 

So my first question is, why doesn't  v-close-popup work to close when you do await or timeouts?

Second, if I then click on the Add Product button again it will close instantly, but also then trigger another @click execution. 
This tells me the  closeSave.value = true; has been executed, but the v-close-popup hasn't worked

It seems the v-close-popup times out or something if it's not set immediately in synchronous code. 

Is there some other way to trigger the closing of the dialog in this case?
I know I can close it via saveFile.value = false; but I'm not sure that's the correct way?
Doing that also doesn't answer the core question here, as to why v-close-popup is not working as expected after an await/timeout.

I have no errors or warnings in the console as to why the above does not work. 

Thanks!

**Comment by rstoenescu** [comment, marked as answer]
Hi,

Your usage of v-close-popup is incorrect. The supplied value for the vue directive should never be a function.

<img width="1000" alt="image" src="https://github.com/quasarframework/quasar/assets/4665641/f32fefa8-c680-498a-8b70-3fd71a9dd68a">

Add the logic to the already attached `click` event (saveProduct). Should you want the dialog to be closed with a delay, then do that from within saveProduct() too (I see that you have a v-model on the dialog so change it to `false`).

**Comment by sonicviz** [comment]
Aok, thanks for the info!
